llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp

```diff
@@ -154,6 +195,14 @@ static bool selectFnForFPBuiltinCalls(const TargetLibraryInfo &TLI,
     }
   }
 
+  // We don't have implementation for CUDA approximate precision builtins.
+  // Lets map them on NVPTX intrinsics. If no appropriate intrinsics are known
+  // - skip to replaceWithAltMathFunction.
+  if (T.isNVPTX() && BuiltinCall.getRequiredAccuracy().value() == 3.0) {
```

Comment:
Looking at the PTX ISA documentation (https://docs.nvidia.com/cuda/parallel-thread-execution/#floating-point-instructions-div), the accuracy varies based on the operation. For example, div has 2 ULP error, and sqrt has 1 ULP error (if I haven't botched my conversion characteristics). The other functions you mention in the above comment have much, much higher ULPs in general.

So I think it makes sense to let the check for replacement vary the needed accuracy on a per-function basis.

Reply:
Thanks! I have applied the comment in this way: check here for `== 3.0` was replaced to `!= 0.5` and accuracy parameter was moved to be a function parameter. Inside of the function I've added a check of accuracy for `fdiv == 2.5` (or for `sqrt == 3.0`), then try to replace `fpbuiltin` with `approx` nvvm intrinsic, otherwise function does nothing expecting `fpbuiltin` intrinsics to be handled by alt math library. May be it makes sense to check, if accuracy for `fdiv >= 2.0` (for `sqrt >= 1.0`) and then to replace with such `fpbuiltin`s with `approx` nvvm intrinsics,, but I'm not sure if it's a good general solution, may be it's better to get a test failure later on during approximate/fast builtins feature development, then silently replace them here.